frontend/src/metabase-lib/v1/expressions/compiler.ts

```diff
@@ -0,0 +1,82 @@
+import * as Lib from "metabase-lib";
+import type Database from "metabase-lib/v1/metadata/Database";
+import type { Expression } from "metabase-types/api";
+
+import { type ExpressionError, renderError } from "./errors";
+import { resolverPass } from "./field-resolver";
+import {
+  adjustBooleans,
+  adjustCaseOrIf,
+  adjustMultiArgOptions,
+  adjustOffset,
+  adjustOptions,
+  applyPasses,
+} from "./passes";
+import { compile, lexify, parse } from "./pratt";
+import type { ClauseType, StartRule } from "./types";
+
+export type CompileResult<S extends StartRule> =
+  | {
+      error: ExpressionError;
+      expression: null;
+      expressionClause: null;
+    }
+  | {
+      error: null;
+      expression: Expression;
+      expressionClause: ClauseType<S>;
+    };
+
+export function compileExpression<S extends StartRule>({
+  source,
+  startRule,
+  query,
+  stageIndex,
+  database,
+  resolve: shouldResolve = true,
+}: {
+  source: string;
+  startRule: S;
+  query: Lib.Query;
+  stageIndex: number;
+  database?: Database | null;
+  resolve?: boolean;
+}): CompileResult<S> {
+  try {
+    const tokens = lexify(source);
+    const { root } = parse(tokens, { throwOnError: true });
+    const compiled = compile(root);
+    const expression = applyPasses(compiled, [
+      adjustOptions,
+      adjustOffset,
+      adjustCaseOrIf,
+      adjustMultiArgOptions,
+      shouldResolve &&
+        resolverPass({
+          database,
+          query,
+          stageIndex,
+          startRule,
+        }),
+      adjustBooleans,
+    ]);
+
+    const expressionClause = Lib.expressionClauseForLegacyExpression(
+      query,
+      stageIndex,
+      expression,
+    ) as ClauseType<S>;
```

Comment:
This is wrong. In all cases it should return `Lib.ExpressionClause`. `Lib.filter` etc support expression clauses. If some places do not, types should be updated to allow `Lib.ExpressionClause`.

Reply:
Ok I've fixed this but this made it necessary to cast `Lib.FilterClause` and `Lib.AggregationClause` to `Lib.ExpressionClause` in some cases (see cf40f53f67eccb0e1335c4bb4d6a20348eba98db).

Maybe we should add a `Lib.isExpressionClause` helper so we can at least verify that the right clause is being passed?
Or we should somehow encode in the type system that these are interchangeable?